File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/role-has-required-aria-props.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getElementType } from "../../utils/get-element-type.js";
import { getJsxPropStringValue } from "../../utils/get-jsx-prop-string-value.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";

const buildMessage = (role: string, missingProps: ReadonlyArray<string>): string =>
  `Screen reader users can't tell the state of this \`${role}\` without its required ARIA props, so add \`${missingProps.join("`, `")}\`.`;

// Mirrors OXC's `ROLE_TO_REQUIRED_ARIA_PROPS`.
const ROLE_REQUIRED_PROPS: ReadonlyMap<string, ReadonlyArray<string>> = new Map([
  ["checkbox", ["aria-checked"]],
  ["combobox", ["aria-controls", "aria-expanded"]],
  ["heading", ["aria-level"]],
  ["menuitemcheckbox", ["aria-checked"]],
  ["menuitemradio", ["aria-checked"]],
  ["meter", ["aria-valuenow"]],
  ["option", ["aria-selected"]],
  ["radio", ["aria-checked"]],
  ["scrollbar", ["aria-controls", "aria-valuenow"]],
  ["slider", ["aria-valuenow"]],
  ["switch", ["aria-checked"]],
]);

const NATIVE_CHECKED_STATE_PROP = "aria-checked";

// A native `<input type="checkbox|radio">` with a `checked`/`defaultChecked`
// binding maps its DOM checked state into the accessibility tree intrinsically,
// so an overriding role (e.g. `switch`) doesn't also need an explicit
// `aria-checked`. A custom `<div role="switch">` has no such intrinsic state.
const suppliesNativeCheckedState = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  settings: Readonly<Record<string, unknown>> | undefined,
): boolean => {
  if (getElementType(node, settings) !== "input") return false;
  const typeAttribute = hasJsxPropIgnoreCase(node.attributes, "type");
  const inputType = typeAttribute ? getJsxPropStringValue(typeAttribute) : null;
  if (inputType !== "checkbox" && inputType !== "radio") return false;
  return (
    Boolean(hasJsxPropIgnoreCase(node.attributes, "checked")) ||
    Boolean(hasJsxPropIgnoreCase(node.attributes, "defaultChecked"))
  );
};

// Port of `oxc_linter::rules::jsx_a11y::role_has_required_aria_props`.
export const roleHasRequiredAriaProps = defineRule({
  id: "role-has-required-aria-props",
  title: "Role missing required ARIA props",
  tags: ["react-jsx-only"],
  severity: "error",
  recommendation:
    "Add every required `aria-*` attribute so assistive tech can expose the role's state correctly.",
  category: "Accessibility",
  create: (context) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      const roleAttribute = hasJsxPropIgnoreCase(node.attributes, "role");
      if (!roleAttribute) return;
      const roleValue = getJsxPropStringValue(roleAttribute);
      if (roleValue === null) return;
      const roles = roleValue.split(/\s+/).filter((token) => token.length > 0);
      const nativeCheckedState = suppliesNativeCheckedState(node, context.settings);
      for (const role of roles) {
        const required = ROLE_REQUIRED_PROPS.get(role);
        if (!required) continue;
        const missing = required.filter((property) => {
          if (property === NATIVE_CHECKED_STATE_PROP && nativeCheckedState) return false;
          return !hasJsxPropIgnoreCase(node.attributes, property);
        });
        if (missing.length > 0) {
          context.report({ node: roleAttribute, message: buildMessage(role, missing) });
        }
      }
    },
  }),
});

File: native/oxlint/rules/role-has-required-aria-props.rs
use oxc_ast::{
    AstKind,
    ast::{JSXAttributeItem, JSXAttributeValue},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode, context::LintContext, globals::HTML_TAG, rule::Rule, utils::has_jsx_prop_ignore_case,
};

const ROLE_REQUIRED_PROPS: &[(&str, &[&str])] = &[
    ("checkbox", &["aria-checked"]),
    ("combobox", &["aria-controls", "aria-expanded"]),
    ("heading", &["aria-level"]),
    ("menuitemcheckbox", &["aria-checked"]),
    ("menuitemradio", &["aria-checked"]),
    ("meter", &["aria-valuenow"]),
    ("option", &["aria-selected"]),
    ("radio", &["aria-checked"]),
    ("scrollbar", &["aria-controls", "aria-valuenow"]),
    ("slider", &["aria-valuenow"]),
    ("switch", &["aria-checked"]),
];

#[derive(Debug, Default, Clone)]
pub struct RoleHasRequiredAriaProps;

declare_oxc_lint!(
    /// Require every ARIA property required by an explicit role.
    RoleHasRequiredAriaProps,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require every ARIA property required by an explicit role.",
);

impl Rule for RoleHasRequiredAriaProps {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        let should_use_curated_behavior = should_use_curated_port_behavior(ctx);
        if should_use_curated_behavior && is_local_test_scaffold_jsx(node, ctx) {
            return;
        }
        let element_type = resolve_configured_jsx_element_type(opening_element, ctx);
        if !HTML_TAG.contains(element_type.as_str()) {
            return;
        }
        let Some(JSXAttributeItem::Attribute(role_attribute)) =
            has_jsx_prop_ignore_case(opening_element, "role")
        else {
            return;
        };
        let Some(role_candidates) = get_static_jsx_attribute_string_values(role_attribute, ctx)
        else {
            return;
        };
        let mut roles = Vec::new();
        for candidate in role_candidates {
            for role in candidate.split_whitespace() {
                if !roles.iter().any(|existing_role| existing_role == role) {
                    roles.push(role.to_string());
                }
            }
        }
        for role in roles {
            let Some((_, required_properties)) = ROLE_REQUIRED_PROPS
                .iter()
                .find(|(required_role, _)| *required_role == role.as_str())
            else {
                continue;
            };
            let missing_properties = required_properties
                .iter()
                .copied()
                .filter(|property| {
                    !(should_use_curated_behavior
                        && supplies_native_aria_property(opening_element, &element_type, property))
                        && has_jsx_prop_ignore_case(opening_element, property).is_none()
                })
                .collect::<Vec<_>>();
            if missing_properties.is_empty() {
                continue;
            }
            let message = format!(
                "Screen reader users can't tell the state of this `{role}` without its required ARIA props, so add `{}`.",
                missing_properties.join("`, `")
            );
            ctx.diagnostic(OxcDiagnostic::warn(message).with_label(role_attribute.span));
        }
    }
}

fn supplies_native_aria_property(
    opening_element: &oxc_ast::ast::JSXOpeningElement<'_>,
    element_type: &str,
    property: &str,
) -> bool {
    if property == "aria-level" && matches!(element_type, "h1" | "h2" | "h3" | "h4" | "h5" | "h6") {
        return true;
    }
    if element_type == "select" {
        return matches!(property, "aria-controls" | "aria-expanded");
    }
    if property == "aria-selected" && element_type == "option" {
        return true;
    }
    if element_type != "input" {
        return false;
    }
    let input_type = has_jsx_prop_ignore_case(opening_element, "type")
        .and_then(|attribute| match attribute {
            JSXAttributeItem::Attribute(attribute) => attribute.value.as_ref(),
            JSXAttributeItem::SpreadAttribute(_) => None,
        })
        .and_then(|value| match value {
            JSXAttributeValue::StringLiteral(string_literal) => Some(string_literal.value.as_str()),
            _ => None,
        });
    if property == "aria-checked" {
        return matches!(input_type, Some("checkbox" | "radio"));
    }
    matches!(
        (input_type, property),
        (
            Some("range"),
            "aria-valuenow" | "aria-valuemin" | "aria-valuemax"
        )
    )
}
